pilot/pkg/model/context.go

```diff
@@ -119,6 +120,226 @@ var (
 	istioVersionRegexp = regexp.MustCompile(`^([1-9]+)\.([0-9]+)(\.([0-9]+))?`)
 )
 
+// StringList is a list that will be marshaled to a comma separate string in Json
+type StringList []string
+
+func (l StringList) MarshalJSON() ([]byte, error) {
+	if l == nil {
+		return nil, nil
+	}
+	return []byte(`"` + strings.Join(l, ",") + `"`), nil
+}
+
+func (l *StringList) UnmarshalJSON(data []byte) error {
+	if len(data) == 0 || string(data) == `""` {
+		*l = []string{}
+	} else {
+		*l = strings.Split(string(data[1:len(data)-1]), ",")
+	}
+	return nil
+}
+
+// PodPort describes a mapping of port name to port number. Generally, this is just the definition of
+// a port in Kubernetes, but without depending on Kubernetes api.
+type PodPort struct {
+	// If specified, this must be an IANA_SVC_NAME and unique within the pod. Each
+	// named port in a pod must have a unique name. Name for the port that can be
+	// referred to by services.
+	// +optional
+	Name string `json:"name,omitempty"`
+	// Number of port to expose on the pod's IP address.
+	// This must be a valid port number, 0 < x < 65536.
+	ContainerPort int `json:"containerPort"`
+	// Name of the protocol
+	Protocol string `json:"protocol"`
+}
+
+// PodPortList defines a list of PodPort's that is serialized as a string
+// This is for legacy reasons, where proper JSON was not supported and was written as a string
+type PodPortList []PodPort
+
+func (l PodPortList) MarshalJSON() ([]byte, error) {
+	if l == nil {
+		return nil, nil
+	}
+	b, err := json.Marshal([]PodPort(l))
+	if err != nil {
+		return nil, err
+	}
+	b = bytes.ReplaceAll(b, []byte{'"'}, []byte{'\\', '"'})
+	out := append([]byte{'"'}, b...)
+	out = append(out, '"')
+	return out, nil
+}
+
+func (l *PodPortList) UnmarshalJSON(data []byte) error {
+	var pl []PodPort
+	pls, err := strconv.Unquote(string(data))
+	if err != nil {
+		return nil
+	}
+	if err := json.Unmarshal([]byte(pls), &pl); err != nil {
+		return err
+	}
+	*l = pl
+	return nil
+}
+
+// NodeMetadata defines the metadata associated with a proxy
+// Fields should not be assumed to exist on the proxy, especially newly added fields which will not exist
+// on older versions.
+// The JSON field names should never change, as they are needed for backward compatibility with older proxies
+type NodeMetadata struct {
+	// IstioVersion specifies the Istio version associated with the proxy
+	IstioVersion string `json:"ISTIO_VERSION,omitempty"`
```

Comment:
@howardjohn Older versions of proxies have the proxy version in "ISTIO_PROXY_VERSION" - was it changed to use "ISTIO_VERSION"?

Here is metdata dump of older (prior to 1.3)

```
     "metadata": {
      "INTERCEPTION_MODE": "REDIRECT",
      "CONFIG_NAMESPACE": "service-mesh",
      "WORKLOAD_NAME": "istio-ordering-66d468b7fb",
      "TLS_CLIENT_ROOT_CERT": "/client-certs/ca.pem",
      "sidecar.istio.io/inject": "true",
      "istio_app": "test",
      "madkub.sam.sfdc.net/allcerts": "{\"certreqs\":[{\"name\":\"tls-client-cert\",\"san\":null,\"role\":\"service-mesh.istio-ordering\",\"cert-type\":\"client\",\"kingdom\":\"prd\",\"superpod\":\"None\"},{\"name\":\"tls-server-cert\",\"san\":null,\"role\":\"service-mesh.istio-ordering\",\"cert-type\":\"server\",\"kingdom\":\"prd\",\"superpod\":\"None\"}]}",
      "POD_NAME": "istio-ordering-66d468b7fb-q5hnn",
      "istio": "sidecar",
      **"ISTIO_PROXY_VERSION": "1.1.3",**
      "sidecar.istio.io/statsInclusionPrefixes": "access_log_file,cluster,cluster_manager,control_plane,http,http2,http_mixer_filter,listener,listener_manager,redis,runtime,server,stats,tcp,tcp_mixer_filter,tracing",
      "INCLUDE_INBOUND_PORTS": "7443,15372",
      "NAMESPACE": "service-mesh",
      "EXCHANGE_KEYS": "NAME,NAMESPACE,INSTANCE_IPS,LABELS,OWNER,PLATFORM_METADATA,WORKLOAD_NAME,CANONICAL_TELEMETRY_SERVICE,MESH_ID,SERVICE_ACCOUNT",
      "TLS_SERVER_CERT_CHAIN": "/server-certs/server/certificates/server.pem",
      "INSTANCE_IPS": "10.251.130.102,fe80::42:aff:fefb:8266",
      "TLS_CLIENT_KEY": "/client-certs/client/keys/client-key.pem",
      "pod-template-hash": "2280246396",
      "POD_PORTS": "[\n    {\"name\":\"grpc-svc\",\"containerPort\":7443,\"protocol\":\"TCP\"}\n    ,{\"name\":\"scone-mgmt\",\"containerPort\":15372,\"protocol\":\"TCP\"}\n]",
      "TLS_SERVER_ROOT_CERT": "/server-certs/ca.pem",
      "SERVICE_ACCOUNT": "default",

      "TLS_CLIENT_CERT_CHAIN": "/client-certs/client/certificates/client.pem",
      "CLUSTER_ID": "Kubernetes",

}
```